src/spec.c

```diff
@@ -81,6 +81,55 @@ static void setMemoryInfo(RedisModuleCtx *ctx) {
   RedisModule_FreeServerInfo(ctx, info);
 }
 
+#define MILLISECOND_IN_ONE_SECOND 1000
+#define NANOSECOND_IN_ONE_MILLISECOND 1000000
+
+static inline timespec timespecFromMilliseconds(int64_t totalMilliseconds) {
+  timespec result = {.tv_sec = 0, .tv_nsec = 0};
+  if (totalMilliseconds > 0) {
+    result.tv_sec = totalMilliseconds / MILLISECOND_IN_ONE_SECOND;
+    result.tv_nsec = (totalMilliseconds % MILLISECOND_IN_ONE_SECOND) * NANOSECOND_IN_ONE_MILLISECOND;
+  }
+  return result;
+}
+
+static inline t_expirationTimePoint getDocExpirationTime(RedisModuleCtx* ctx, RedisModuleString *keystr) {
+  timespec zero = {.tv_sec = 0, .tv_nsec = 0};
+  RedisModuleCallReply *rep = RedisModule_Call(ctx, "PEXPIRETIME", "s", keystr);
+  if (rep == NULL || RedisModule_CallReplyType(rep) != REDISMODULE_REPLY_INTEGER) {
+    size_t length = 0;
+    const char* error = rep ? RedisModule_CallReplyStringPtr(rep, &length) : "unknown error";
+    RedisModule_Log(ctx, "warning", "Error calling PEXPIRETIME, type: %d, error(%d): %s", rep ? RedisModule_CallReplyType(rep) : REDISMODULE_REPLY_UNKNOWN, errno, error);
+    return zero;
+  }
+
+  const int64_t totalMilliseconds = RedisModule_CallReplyInteger(rep);
+  t_expirationTimePoint result = timespecFromMilliseconds(totalMilliseconds);
+  RedisModule_FreeCallReply(rep);
+  return result;
+}
+
+static inline FieldExpiration* callHashFieldExpirationTime(RedisModuleCtx* ctx, RedisModuleString *keystr, RedisModuleString** fields) {
+  size_t vectorElementCount = array_len(fields);
+  long long fieldCount = vectorElementCount;
+  RedisModuleCallReply *rep = RedisModule_Call(ctx, "HPEXPIRETIME", "sclv", keystr, "FIELDS", fieldCount, fields, vectorElementCount);
+  if (rep == NULL || RedisModule_CallReplyType(rep) != REDISMODULE_REPLY_ARRAY) {
+    return NULL;
+  }
+  size_t sz = RedisModule_CallReplyLength(rep);
+  arrayof(FieldExpiration) result = NULL;
+  for (t_fieldIndex i = 0; i < sz; i++) {
+    const int64_t milliseconds = RedisModule_CallReplyInteger(RedisModule_CallReplyArrayElement(rep, i));
+    if (milliseconds <= 0) {
```

Comment:
Isn't <0 mean no expiration and ==0 mean just expired? Is it OK to treat both cases the same?

Reply:
HPEXPIRETIME returns a unix epoch timestamp for every field, or -1/-2 as errors.
If a field just expired but we didn't get a notification for it then we should still get a unix epoch timestamp for it.

`[Array reply](https://redis.io/docs/latest/develop/reference/protocol-spec#arrays). For each field:
        [Integer reply](https://redis.io/docs/latest/develop/reference/protocol-spec#integers): -2 if no such field exists in the provided hash key, or the provided key does not exist.
        [Integer reply](https://redis.io/docs/latest/develop/reference/protocol-spec#integers): -1 if the field exists but has no associated expiration set.
        [Integer reply](https://redis.io/docs/latest/develop/reference/protocol-spec#integers): the expiration (Unix timestamp) in seconds.`

Reply:
From the code it seems that "just expired" docs are rounded up to 0. I don't think it's the same case

Reply:
Where did you see it indicating an invalid timestamp? I would assume invalid TTL input is handled when it is set

Reply:
0 in the context of unix epoch time means the following date:
Thursday, January 1, 1970 12:00:00 AM

The expiration time the user set is always relative to the current point in time, i.e now.
So it is impossible for a user to get the epoch to be 0.
And in the context of "now", 0 would be an invalid timestamp since our product didn't exist before that time.

To make it clearer in the code I'll remove the = from the if.